# Objective
* Build a Model which identify the bad customer for loan disburement
* Yes = loan Reject ( Bad Customers ) , No =loan approved ( Good Customers )

In [47]:
################# Data Analysis & Calculation #################
import numpy as np 
import pandas as pd  

import os

################# ignore Warning #################
import warnings
warnings.filterwarnings("ignore")

################# visualization #################
import matplotlib.pyplot as plt 
import seaborn as sns 

# 1. Data Importing

In [48]:
df = pd.read_csv(r"C:\Users\KaushalDhumal\Downloads\ML\Bank_Loan.csv")
df.head()

Loan_ID  Age Gender Married  Dependents     Education Self_Employed  \
0  LP001002   43   Male      No           3      Graduate            No   
1  LP001003   57   Male     Yes           2      Graduate            No   
2  LP001005   25   Male     Yes           3      Graduate           Yes   
3  LP001006   44   Male     Yes           1  Not Graduate            No   
4  LP001008   58   Male      No           1      Graduate            No   

   ApplicantIncome  LoanAmount Previous_Loan_Taken  Cibil_Score Property_Area  \
0            54005      334113                 Yes          668         Urban   
1            28817      736026                 Yes          684         Rural   
2            99939      712432                 Yes          788         Urban   
3            38470      267161                 Yes          888         Urban   
4            28991      402474                 Yes          703         Urban   

  Customer_Bandwith  Tenure Loan_Status  
0              Good      43          No  
1              Good      13         Yes  
2              Good      10          No  
3              Good       9          No  
4              Good      11          No

# 2. Data Pre-Processing
* Data Type and conversion
* Identifying and Treating missing value

### A) Data Type and conversion

In [49]:
df = df.drop('Loan_ID', axis=1)
df.head()

Age Gender Married  Dependents     Education Self_Employed  \
0   43   Male      No           3      Graduate            No   
1   57   Male     Yes           2      Graduate            No   
2   25   Male     Yes           3      Graduate           Yes   
3   44   Male     Yes           1  Not Graduate            No   
4   58   Male      No           1      Graduate            No   

   ApplicantIncome  LoanAmount Previous_Loan_Taken  Cibil_Score Property_Area  \
0            54005      334113                 Yes          668         Urban   
1            28817      736026                 Yes          684         Rural   
2            99939      712432                 Yes          788         Urban   
3            38470      267161                 Yes          888         Urban   
4            28991      402474                 Yes          703         Urban   

  Customer_Bandwith  Tenure Loan_Status  
0              Good      43          No  
1              Good      13         Yes  
2              Good      10          No  
3              Good       9          No  
4              Good      11          No

In [50]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 981 entries, 0 to 980
Data columns (total 14 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Age                  981 non-null    int64 
 1   Gender               957 non-null    object
 2   Married              978 non-null    object
 3   Dependents           981 non-null    int64 
 4   Education            981 non-null    object
 5   Self_Employed        926 non-null    object
 6   ApplicantIncome      981 non-null    int64 
 7   LoanAmount           981 non-null    int64 
 8   Previous_Loan_Taken  981 non-null    object
 9   Cibil_Score          981 non-null    int64 
 10  Property_Area        981 non-null    object
 11  Customer_Bandwith    981 non-null    object
 12  Tenure               981 non-null    int64 
 13  Loan_Status          981 non-null    object
dtypes: int64(6), object(8)
memory usage: 107.4+ KB


### B) Identifying and Treating missing values

In [51]:
df.isnull().sum()

Age                     0
Gender                 24
Married                 3
Dependents              0
Education               0
Self_Employed          55
ApplicantIncome         0
LoanAmount              0
Previous_Loan_Taken     0
Cibil_Score             0
Property_Area           0
Customer_Bandwith       0
Tenure                  0
Loan_Status             0
dtype: int64

#### The 3 columns which are hvaing missing values are categorical , hence we can treat null values by Mode.

In [52]:
df.Gender.value_counts(dropna=False)

Gender
Male      775
Female    182
NaN        24
Name: count, dtype: int64

In [53]:
for i in ['Gender', 'Married', 'Self_Employed']:
    df[i].fillna(df[i].mode()[0], inplace=True)

In [54]:
df.isnull().sum()

Age                    0
Gender                 0
Married                0
Dependents             0
Education              0
Self_Employed          0
ApplicantIncome        0
LoanAmount             0
Previous_Loan_Taken    0
Cibil_Score            0
Property_Area          0
Customer_Bandwith      0
Tenure                 0
Loan_Status            0
dtype: int64

# 3. Data Visualization

In [55]:
df.Loan_Status.value_counts(normalize=True)

Loan_Status
No     0.72579
Yes    0.27421
Name: proportion, dtype: float64

In [56]:
import plotly.express as px

fig = px.pie(df, names='Loan_Status', color='Loan_Status',
            color_discrete_map={'Yes':'Yellow',
                               'No':'Blue'})
fig.show()

# 4. Breaking Data into 2 parts
* Quantitative variable
* Qualitative variable

In [57]:
df.head()

Age Gender Married  Dependents     Education Self_Employed  \
0   43   Male      No           3      Graduate            No   
1   57   Male     Yes           2      Graduate            No   
2   25   Male     Yes           3      Graduate           Yes   
3   44   Male     Yes           1  Not Graduate            No   
4   58   Male      No           1      Graduate            No   

   ApplicantIncome  LoanAmount Previous_Loan_Taken  Cibil_Score Property_Area  \
0            54005      334113                 Yes          668         Urban   
1            28817      736026                 Yes          684         Rural   
2            99939      712432                 Yes          788         Urban   
3            38470      267161                 Yes          888         Urban   
4            28991      402474                 Yes          703         Urban   

  Customer_Bandwith  Tenure Loan_Status  
0              Good      43          No  
1              Good      13         Yes  
2              Good      10          No  
3              Good       9          No  
4              Good      11          No

In [58]:
Quantitative_var = df.select_dtypes(include=[np.number])
Quantitative_var.head()

Age  Dependents  ApplicantIncome  LoanAmount  Cibil_Score  Tenure
0   43           3            54005      334113          668      43
1   57           2            28817      736026          684      13
2   25           3            99939      712432          788      10
3   44           1            38470      267161          888       9
4   58           1            28991      402474          703      11

In [59]:
Qualitative_var = df.select_dtypes(include=['object'])
Qualitative_var.head()

Gender Married     Education Self_Employed Previous_Loan_Taken  \
0   Male      No      Graduate            No                 Yes   
1   Male     Yes      Graduate            No                 Yes   
2   Male     Yes      Graduate           Yes                 Yes   
3   Male     Yes  Not Graduate            No                 Yes   
4   Male      No      Graduate            No                 Yes   

  Property_Area Customer_Bandwith Loan_Status  
0         Urban              Good          No  
1         Rural              Good         Yes  
2         Urban              Good          No  
3         Urban              Good          No  
4         Urban              Good          No

## Converting Qualitative variable into number by-Label Encoder

In [60]:
from sklearn.preprocessing import LabelEncoder

Qualitative_var = Qualitative_var.apply(LabelEncoder().fit_transform)
Qualitative_var.head()

Gender  Married  Education  Self_Employed  Previous_Loan_Taken  \
0       1        0          0              0                    1   
1       1        1          0              0                    1   
2       1        1          0              1                    1   
3       1        1          1              0                    1   
4       1        0          0              0                    1   

   Property_Area  Customer_Bandwith  Loan_Status  
0              2                  1            0  
1              0                  1            1  
2              2                  1            0  
3              2                  1            0  
4              2                  1            0

## Combining the Data set

In [61]:
data = pd.concat([Quantitative_var, Qualitative_var], axis=1)
data.head()

Age  Dependents  ApplicantIncome  LoanAmount  Cibil_Score  Tenure  Gender  \
0   43           3            54005      334113          668      43       1   
1   57           2            28817      736026          684      13       1   
2   25           3            99939      712432          788      10       1   
3   44           1            38470      267161          888       9       1   
4   58           1            28991      402474          703      11       1   

   Married  Education  Self_Employed  Previous_Loan_Taken  Property_Area  \
0        0          0              0                    1              2   
1        1          0              0                    1              0   
2        1          0              1                    1              2   
3        1          1              0                    1              2   
4        0          0              0                    1              2   

   Customer_Bandwith  Loan_Status  
0                  1            0  
1                  1            1  
2                  1            0  
3                  1            0  
4                  1            0

# 5. Data Partition

In [62]:
from sklearn.model_selection import train_test_split

x = data.drop('Loan_Status', axis=1)
y = data[['Loan_Status']]

X_train,X_test,y_train,y_test = train_test_split(x,y,train_size=0.7,random_state=126)

In [63]:
train = pd.concat([X_train,y_train], axis=1)
train.head()

Age  Dependents  ApplicantIncome  LoanAmount  Cibil_Score  Tenure  \
938   45           4            53985      881765          709      34   
292   53           0            32434      994669          883      37   
457   21           5            34229      866491          467       9   
158   46           0            75349      814093          615       6   
488   41           3            61916      496945          857      16   

     Gender  Married  Education  Self_Employed  Previous_Loan_Taken  \
938       1        1          0              0                    1   
292       1        1          0              0                    1   
457       1        1          0              0                    1   
158       1        0          0              0                    1   
488       1        1          0              1                    1   

     Property_Area  Customer_Bandwith  Loan_Status  
938              0                  1            0  
292              2                  1            0  
457              2                  2            1  
158              0                  2            0  
488              1                  1            0

# 6. Model Building

In [64]:
from sklearn.feature_selection import SequentialFeatureSelector as sfs
from sklearn.linear_model import LogisticRegression

logreg = LogisticRegression(random_state=126)
model = sfs(logreg, n_features_to_select=6, direction='forward', scoring='accuracy')
model.fit(X_train,y_train)

SequentialFeatureSelector(estimator=LogisticRegression(random_state=126),
                          n_features_to_select=6, scoring='accuracy')

In [65]:
model.get_feature_names_out()

array(['Dependents', 'Cibil_Score', 'Gender', 'Married', 'Self_Employed',
       'Previous_Loan_Taken'], dtype=object)

In [66]:
X_train = X_train.loc[:,['Dependents', 'Cibil_Score', 'Gender', 'Married', 'Self_Employed','Previous_Loan_Taken']]
X_train.head()

Dependents  Cibil_Score  Gender  Married  Self_Employed  \
938           4          709       1        1              0   
292           0          883       1        1              0   
457           5          467       1        1              0   
158           0          615       1        0              0   
488           3          857       1        1              1   

     Previous_Loan_Taken  
938                    1  
292                    1  
457                    1  
158                    1  
488                    1

In [67]:
from sklearn.linear_model import LogisticRegression

logreg = LogisticRegression()
model2 = logreg.fit(X_train,y_train)

In [68]:
model2.intercept_

array([8.5242455])

In [69]:
model2.coef_

array([[ 0.03893837, -0.00963334, -0.12270733, -0.60851403,  0.10403524,
        -3.27869845]])

In [70]:
model2.feature_names_in_

array(['Dependents', 'Cibil_Score', 'Gender', 'Married', 'Self_Employed',
       'Previous_Loan_Taken'], dtype=object)

# 7. Predictions on Train Dataset

In [71]:
train = pd.concat([X_train,y_train], axis=1)
train.head()

Dependents  Cibil_Score  Gender  Married  Self_Employed  \
938           4          709       1        1              0   
292           0          883       1        1              0   
457           5          467       1        1              0   
158           0          615       1        0              0   
488           3          857       1        1              1   

     Previous_Loan_Taken  Loan_Status  
938                    1            0  
292                    1            0  
457                    1            1  
158                    1            0  
488                    1            0

In [74]:
train['Probability_bad'] = model2.predict_proba(X_train)[:,1]
train.head(10)

Dependents  Cibil_Score  Gender  Married  Self_Employed  \
938           4          709       1        1              0   
292           0          883       1        1              0   
457           5          467       1        1              0   
158           0          615       1        0              0   
488           3          857       1        1              1   
935           2          775       0        1              0   
840           2          672       1        1              0   
257           5          797       1        1              0   
122           1          796       0        0              0   
644           1          770       0        0              0   

     Previous_Loan_Taken  Loan_Status  Probability_bad  
938                    1            0         0.103404  
292                    1            0         0.018130  
457                    1            1         0.552369  
158                    1            0         0.309669  
488                    1            0         0.028732  
935                    1            0         0.060034  
840                    1            0         0.132228  
257                    1            1         0.048857  
122                    0            0         0.709930  
644                    1            0         0.105913

In [75]:
train['Predicted'] = np.where(train['Probability_bad'] >= 0.7,1,0)
train.head(10)

Dependents  Cibil_Score  Gender  Married  Self_Employed  \
938           4          709       1        1              0   
292           0          883       1        1              0   
457           5          467       1        1              0   
158           0          615       1        0              0   
488           3          857       1        1              1   
935           2          775       0        1              0   
840           2          672       1        1              0   
257           5          797       1        1              0   
122           1          796       0        0              0   
644           1          770       0        0              0   

     Previous_Loan_Taken  Loan_Status  Probability_bad  Predicted  
938                    1            0         0.103404          0  
292                    1            0         0.018130          0  
457                    1            1         0.552369          0  
158                    1            0         0.309669          0  
488                    1            0         0.028732          0  
935                    1            0         0.060034          0  
840                    1            0         0.132228          0  
257                    1            1         0.048857          0  
122                    0            0         0.709930          1  
644                    1            0         0.105913          0

# 8. Model Performance Metrics

In [76]:
from sklearn.metrics import confusion_matrix

matrix = confusion_matrix(train['Predicted'], train['Loan_Status'])
matrix

array([[486,  81],
       [ 11, 108]], dtype=int64)

In [79]:
Accuracy_train = ((486+108)/(686)*100)
Accuracy_train

86.58892128279884

###  Accuracy of Model is 86.59%, So we can say Model is good fit on train dataset

In [82]:
from sklearn.metrics import classification_report

print(classification_report(train['Loan_Status'], train['Predicted']))

              precision    recall  f1-score   support

           0       0.86      0.98      0.91       497
           1       0.91      0.57      0.70       189

    accuracy                           0.87       686
   macro avg       0.88      0.77      0.81       686
weighted avg       0.87      0.87      0.86       686



### 
* Accuracy of Model is 87%,Hence Model is good Fit on Train dataset
* Accuracy of Bad Customer Capture by Model is 57% ( Sensitivity )
* Accuracy of Good Customer Capture by Model is 98% (Specificity)

# 9. Predictions on Test dataset

In [85]:
X_test = X_test.loc[:,['Dependents', 'Cibil_Score', 'Gender', 'Married', 'Self_Employed','Previous_Loan_Taken']]
X_test.head()

Dependents  Cibil_Score  Gender  Married  Self_Employed  \
45            0          609       0        0              0   
735           4          800       1        1              0   
756           5          408       1        1              0   
657           4          887       1        0              0   
423           5          538       1        1              0   

     Previous_Loan_Taken  
45                     1  
735                    1  
756                    0  
657                    1  
423                    0

In [86]:
test = pd.concat([X_test,y_test], axis=1)
test.head()

Dependents  Cibil_Score  Gender  Married  Self_Employed  \
45            0          609       0        0              0   
735           4          800       1        1              0   
756           5          408       1        1              0   
657           4          887       1        0              0   
423           5          538       1        1              0   

     Previous_Loan_Taken  Loan_Status  
45                     1            0  
735                    1            0  
756                    0            1  
657                    1            0  
423                    0            1

In [88]:
test['Probability_Bad'] = model2.predict_proba(X_test)[:,1]
test.head()

Dependents  Cibil_Score  Gender  Married  Self_Employed  \
45            0          609       0        0              0   
735           4          800       1        1              0   
756           5          408       1        1              0   
657           4          887       1        0              0   
423           5          538       1        1              0   

     Previous_Loan_Taken  Loan_Status  Probability_Bad  
45                     1            0         0.349518  
735                    1            0         0.045800  
756                    0            1         0.982999  
657                    1            0         0.036750  
423                    0            1         0.942944

In [89]:
test['Predicted'] = np.where(test['Probability_Bad'] >= 0.7,1,0) 
test.head()

Dependents  Cibil_Score  Gender  Married  Self_Employed  \
45            0          609       0        0              0   
735           4          800       1        1              0   
756           5          408       1        1              0   
657           4          887       1        0              0   
423           5          538       1        1              0   

     Previous_Loan_Taken  Loan_Status  Probability_Bad  Predicted  
45                     1            0         0.349518          0  
735                    1            0         0.045800          0  
756                    0            1         0.982999          1  
657                    1            0         0.036750          0  
423                    0            1         0.942944          1

# 10. Model Performance Metrics on Test dataset

In [90]:
from sklearn.metrics import confusion_matrix

matrix = confusion_matrix(test['Predicted'],test['Loan_Status'])
print(matrix)

[[213  37]
 [  2  43]]


In [91]:
Accuracy_test = ((213+43)/(295)*100)
Accuracy_test

86.77966101694915

In [92]:
from sklearn.metrics import classification_report

print(classification_report(test['Loan_Status'],test['Predicted']))

              precision    recall  f1-score   support

           0       0.85      0.99      0.92       215
           1       0.96      0.54      0.69        80

    accuracy                           0.87       295
   macro avg       0.90      0.76      0.80       295
weighted avg       0.88      0.87      0.85       295



## 
* Accuracy of Model is 87%,Hence Model Performance is good Fit on test Data
* Accuracy of Bad Customer Capture by Model is 54% ( Sensitivity )
* Accuracy of Good Customer Capture by Model is 99%

# Business/Client Submission
* For future Prediction below 5 variable data is mandatory

* Note :- If anyone is not present output will not be reliable

In [94]:
x = list(X_train.columns)
x.insert(0,'Intercept')
model_values = pd.DataFrame(np.concatenate((model2.intercept_.tolist(),model2.coef_.flatten())), index=x, columns=['Coefficient'])
model_values

Coefficient
Intercept               8.524245
Dependents              0.038938
Cibil_Score            -0.009633
Gender                 -0.122707
Married                -0.608514
Self_Employed           0.104035
Previous_Loan_Taken    -3.278698

# 
y = 8.524245 - 0.009633 * Cibil_Score + 0.038938 * Dependents - 0.122707 * Gender -0.608514 * Married + 0.104035 * Self_Employed - 3.278698 * Previous_Loan_Taken 

In [95]:
from sklearn.metrics import classification_report

print(classification_report(train['Loan_Status'],train['Predicted']))

              precision    recall  f1-score   support

           0       0.86      0.98      0.91       497
           1       0.91      0.57      0.70       189

    accuracy                           0.87       686
   macro avg       0.88      0.77      0.81       686
weighted avg       0.87      0.87      0.86       686



# Conclusion % Insights

*This project implemented a logistic regression model to predict bank loan approvals. After preprocessing (handling missing values, encoding categorical variables, and normalizing features), the model was trained and evaluated.

## Key Insight:
* Important Features: 'Dependents', 'Cibil_Score', 'Gender', 'Married', 'Self_Employed',
       'Previous_Loan_Taken'
           
* Data Handling: Replacing missing categorical values with mode helped maintain data integrity.

* Class Imbalance: If present, techniques like SMOTE or class weighting can improve model performance.

* Model Performance: While logistic regression worked well, models like Random Forest or XGBoost may yield better accuracy.

* Business Use: Banks can leverage such models to automate loan approvals and improve risk assessment.